libxshmfence can not be built on rv32

Hi, I want to install vim in my qemuriscv32 machine.
I already asked about it and got the answer.
I added `IMAGE_INSTALL_append = " vim"` in local.conf, but the following error is appeared.
```
ERROR: libxshmfence-1.3-r0 do_compile: oe_runmake failed
ERROR: libxshmfence-1.3-r0 do_compile: Execution of '/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/temp/run.do_compile.1897' failed with exit code 1:
make  all-recursive
make[1]: Entering directory '/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/build'
Making all in src
make[2]: Entering directory '/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/build/src'
../riscv32-oe-linux-libtool  --tag=CC   --mode=compile riscv32-oe-linux-gcc    -fstack-protector-strong  -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security --sysroot=/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/recipe-sysroot -DHAVE_CONFIG_H -I. -I../../libxshmfence-1.3/src -I..    -Wall -Wpointer-arith -Wmissing-declarations -Wformat=2 -Wstrict-prototypes -Wmissing-prototypes -Wnested-externs -Wbad-function-cast -Wold-style-definition -Wdeclaration-after-statement -Wunused -Wuninitialized -Wshadow -Wmissing-noreturn -Wmissing-format-attribute -Wredundant-decls -Wlogical-op -Werror=implicit -Werror=nonnull -Werror=init-self -Werror=main -Werror=missing-braces -Werror=sequence-point -Werror=return-type -Werror=trigraphs -Werror=array-bounds -Werror=write-strings -Werror=address -Werror=int-to-pointer-cast -Werror=pointer-to-int-cast -fno-strict-aliasing -O2 -pipe -g -feliminate-unused-debug-types -fmacro-prefix-map=/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0=/usr/src/debug/libxshmfence/1.3-r0                      -fdebug-prefix-map=/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0=/usr/src/debug/libxshmfence/1.3-r0                      -fdebug-prefix-map=/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/recipe-sysroot=                      -fdebug-prefix-map=/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/recipe-sysroot-native=   -fvisibility=hidden -c -o xshmfence_alloc.lo ../../libxshmfence-1.3/src/xshmfence_alloc.c
../riscv32-oe-linux-libtool  --tag=CC   --mode=compile riscv32-oe-linux-gcc    -fstack-protector-strong  -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security --sysroot=/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/recipe-sysroot -DHAVE_CONFIG_H -I. -I../../libxshmfence-1.3/src -I..    -Wall -Wpointer-arith -Wmissing-declarations -Wformat=2 -Wstrict-prototypes -Wmissing-prototypes -Wnested-externs -Wbad-function-cast -Wold-style-definition -Wdeclaration-after-statement -Wunused -Wuninitialized -Wshadow -Wmissing-noreturn -Wmissing-format-attribute -Wredundant-decls -Wlogical-op -Werror=implicit -Werror=nonnull -Werror=init-self -Werror=main -Werror=missing-braces -Werror=sequence-point -Werror=return-type -Werror=trigraphs -Werror=array-bounds -Werror=write-strings -Werror=address -Werror=int-to-pointer-cast -Werror=pointer-to-int-cast -fno-strict-aliasing -O2 -pipe -g -feliminate-unused-debug-types -fmacro-prefix-map=/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0=/usr/src/debug/libxshmfence/1.3-r0                      -fdebug-prefix-map=/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0=/usr/src/debug/libxshmfence/1.3-r0                      -fdebug-prefix-map=/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/recipe-sysroot=                      -fdebug-prefix-map=/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/recipe-sysroot-native=   -fvisibility=hidden -c -o xshmfence_futex.lo ../../libxshmfence-1.3/src/xshmfence_futex.c
riscv32-oe-linux-libtool: compile:  riscv32-oe-linux-gcc -fstack-protector-strong -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security --sysroot=/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/recipe-sysroot -DHAVE_CONFIG_H -I. -I../../libxshmfence-1.3/src -I.. -Wall -Wpointer-arith -Wmissing-declarations -Wformat=2 -Wstrict-prototypes -Wmissing-prototypes -Wnested-externs -Wbad-function-cast -Wold-style-definition -Wdeclaration-after-statement -Wunused -Wuninitialized -Wshadow -Wmissing-noreturn -Wmissing-format-attribute -Wredundant-decls -Wlogical-op -Werror=implicit -Werror=nonnull -Werror=init-self -Werror=main -Werror=missing-braces -Werror=sequence-point -Werror=return-type -Werror=trigraphs -Werror=array-bounds -Werror=write-strings -Werror=address -Werror=int-to-pointer-cast -Werror=pointer-to-int-cast -fno-strict-aliasing -O2 -pipe -g -feliminate-unused-debug-types -fmacro-prefix-map=/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0=/usr/src/debug/libxshmfence/1.3-r0 -fdebug-prefix-map=/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0=/usr/src/debug/libxshmfence/1.3-r0 -fdebug-prefix-map=/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/recipe-sysroot= -fdebug-prefix-map=/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/recipe-sysroot-native= -fvisibility=hidden -c ../../libxshmfence-1.3/src/xshmfence_alloc.c  -fPIC -DPIC -o .libs/xshmfence_alloc.o
riscv32-oe-linux-libtool: compile:  riscv32-oe-linux-gcc -fstack-protector-strong -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security --sysroot=/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/recipe-sysroot -DHAVE_CONFIG_H -I. -I../../libxshmfence-1.3/src -I.. -Wall -Wpointer-arith -Wmissing-declarations -Wformat=2 -Wstrict-prototypes -Wmissing-prototypes -Wnested-externs -Wbad-function-cast -Wold-style-definition -Wdeclaration-after-statement -Wunused -Wuninitialized -Wshadow -Wmissing-noreturn -Wmissing-format-attribute -Wredundant-decls -Wlogical-op -Werror=implicit -Werror=nonnull -Werror=init-self -Werror=main -Werror=missing-braces -Werror=sequence-point -Werror=return-type -Werror=trigraphs -Werror=array-bounds -Werror=write-strings -Werror=address -Werror=int-to-pointer-cast -Werror=pointer-to-int-cast -fno-strict-aliasing -O2 -pipe -g -feliminate-unused-debug-types -fmacro-prefix-map=/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0=/usr/src/debug/libxshmfence/1.3-r0 -fdebug-prefix-map=/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0=/usr/src/debug/libxshmfence/1.3-r0 -fdebug-prefix-map=/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/recipe-sysroot= -fdebug-prefix-map=/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/recipe-sysroot-native= -fvisibility=hidden -c ../../libxshmfence-1.3/src/xshmfence_futex.c  -fPIC -DPIC -o .libs/xshmfence_futex.o
In file included from ../../libxshmfence-1.3/src/xshmfenceint.h:32,
                 from ../../libxshmfence-1.3/src/xshmfence_alloc.c:27:
../../libxshmfence-1.3/src/xshmfence_futex.h: In function 'sys_futex':
../../libxshmfence-1.3/src/xshmfence_futex.h:58:17: error: 'SYS_futex' undeclared (first use in this function); did you mean 'sys_futex'?
   58 |  return syscall(SYS_futex, addr1, op, val1, timeout, addr2, val3);
      |                 ^~~~~~~~~
      |                 sys_futex
../../libxshmfence-1.3/src/xshmfence_futex.h:58:17: note: each undeclared identifier is reported only once for each function it appears in
../../libxshmfence-1.3/src/xshmfence_alloc.c: At top level:
../../libxshmfence-1.3/src/xshmfence_alloc.c:51: warning: "MFD_CLOEXEC" redefined
   51 | #define MFD_CLOEXEC  0x0001U
      |
In file included from /home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/recipe-sysroot/usr/include/bits/mman-linux.h:113,
                 from /home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/recipe-sysroot/usr/include/bits/mman.h:32,
                 from /home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/recipe-sysroot/usr/include/sys/mman.h:41,
                 from ../../libxshmfence-1.3/src/xshmfenceint.h:28,
                 from ../../libxshmfence-1.3/src/xshmfence_alloc.c:27:
/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/recipe-sysroot/usr/include/bits/mman-shared.h:30: note: this is the location of the previous definition
   30 | #  define MFD_CLOEXEC 1U
      |
../../libxshmfence-1.3/src/xshmfence_alloc.c:52: warning: "MFD_ALLOW_SEALING" redefined
   52 | #define MFD_ALLOW_SEALING 0x0002U
      |
In file included from /home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/recipe-sysroot/usr/include/bits/mman-linux.h:113,
                 from /home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/recipe-sysroot/usr/include/bits/mman.h:32,
                 from /home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/recipe-sysroot/usr/include/sys/mman.h:41,
                 from ../../libxshmfence-1.3/src/xshmfenceint.h:28,
                 from ../../libxshmfence-1.3/src/xshmfence_alloc.c:27:
/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/recipe-sysroot/usr/include/bits/mman-shared.h:31: note: this is the location of the previous definition
   31 | #  define MFD_ALLOW_SEALING 2U
      |
Makefile:482: recipe for target 'xshmfence_alloc.lo' failed
make[2]: *** [xshmfence_alloc.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from ../../libxshmfence-1.3/src/xshmfenceint.h:32,
                 from ../../libxshmfence-1.3/src/xshmfence_futex.c:27:
../../libxshmfence-1.3/src/xshmfence_futex.h: In function 'sys_futex':
../../libxshmfence-1.3/src/xshmfence_futex.h:58:17: error: 'SYS_futex' undeclared (first use in this function); did you mean 'sys_futex'?
   58 |  return syscall(SYS_futex, addr1, op, val1, timeout, addr2, val3);
      |                 ^~~~~~~~~
      |                 sys_futex
../../libxshmfence-1.3/src/xshmfence_futex.h:58:17: note: each undeclared identifier is reported only once for each function it appears in
../../libxshmfence-1.3/src/xshmfence_futex.h:59:1: error: control reaches end of non-void function [-Werror=return-type]
   59 | }
      | ^
cc1: some warnings being treated as errors
Makefile:482: recipe for target 'xshmfence_futex.lo' failed
make[2]: *** [xshmfence_futex.lo] Error 1
make[2]: Leaving directory '/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/build/src'
Makefile:496: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/build'
Makefile:405: recipe for target 'all' failed
make: *** [all] Error 2
WARNING: exit code 1 from a shell command.

ERROR: Logfile of failure stored in: /home/ca2020/riscv-yocto/build/tmp-glibc/work/riscv32-oe-linux/libxshmfence/1.3-r0/temp/log.do_compile.1897
```
`ERROR: Task (/home/ca2020/riscv-yocto/openembedded-core/meta/recipes-graphics/xorg-lib/libxshmfence_1.3.bb:do_compile) failed with exit code '1'`

How can I solve this problem?

--- a/recipes-graphics/mesa/mesa/rv32-sys-futex.patch
+++ b/recipes-graphics/mesa/mesa/rv32-sys-futex.patch
@@ -1,0 +1,13 @@
+--- a/src/util/futex.h
++++ b/src/util/futex.h
+@@ -26,6 +26,10 @@
+ 
+ #if defined(HAVE_LINUX_FUTEX_H)
+ 
++#if !defined(__NR_futex) && defined(__riscv) && __riscv_xlen == 32
++# define __NR_futex __NR_futex_time64
++#endif
++
+ #include <limits.h>
+ #include <stdint.h>
+ #include <unistd.h>

--- a/recipes-graphics/mesa/mesa_19.%.bbappend
+++ b/recipes-graphics/mesa/mesa_19.%.bbappend
@@ -1,0 +1,3 @@
+FILESEXTRAPATHS_prepend := "${THISDIR}/${PN}:"
+
+SRC_URI_append_riscv32 = " file://rv32-sys-futex.patch"

--- a/recipes-graphics/xorg-lib/libxshmfence/rv32_sys_futex.patch
+++ b/recipes-graphics/xorg-lib/libxshmfence/rv32_sys_futex.patch
@@ -1,0 +1,13 @@
+--- a/src/xshmfence_futex.h
++++ b/src/xshmfence_futex.h
+@@ -47,6 +47,10 @@ static inline int futex_wait(int32_t *ad
+ 
+ #else
+ 
++#if !defined(__NR_futex) && defined(__riscv) && __riscv_xlen == 32
++# define __NR_futex __NR_futex_time64
++#endif
++
+ #include <stdint.h>
+ #include <values.h>
+ #include <linux/futex.h>

--- a/recipes-graphics/xorg-lib/libxshmfence_1.3.bbappend
+++ b/recipes-graphics/xorg-lib/libxshmfence_1.3.bbappend
@@ -1,0 +1,4 @@
+FILESEXTRAPATHS_prepend := "${THISDIR}/${PN}:"
+
+SRC_URI_append_riscv32 = " file://rv32_sys_futex.patch"
+

